### ISSUE TITLE
model flexibility

### ISSUE THREAD

**Issue by yuchenhou** [yuchenhou/elephant#20]
The latest experiment suggests model's flexibility is very high. The training error almost vanished after 100 epochs. 
Below is the experiment's details.

### model configuration
```json
{
  "embedding_size": 100,
  "hidden_units_formation": [
    100,
    100,
    100
  ],
  "steps": 1000000
}
```
### dataset specs
```json
{
  "name": "MovieLens 100K Dataset",
  "n_users": 1000,
  "n_items": 1700,
  "n_ratings": 100000,
  "rating_values": [1, 2, 3, 4, 5],
  "url": "http://grouplens.org/datasets/movielens/"
}
```
### training log
```
Step #1, avg. loss: 17.59876
Step #100001, epoch #35, avg. loss: 0.59834
Step #200001, epoch #70, avg. loss: 0.11111
Step #300001, epoch #105, avg. loss: 0.02714
Step #400001, epoch #141, avg. loss: 0.01094
Step #500001, epoch #176, avg. loss: 0.00586
Step #600001, epoch #211, avg. loss: 0.00374
Step #700001, epoch #247, avg. loss: 0.00254
Step #800001, epoch #282, avg. loss: 0.00190
Step #900001, epoch #317, avg. loss: 0.00155
```
time cost: 1h 10m

### comments
The noise in the data should be much higher than the training error. There are 2 reasons:

9. data noise > 0.5: We know the rating system's granularity is 1(it only accept integer ratings, or in other words, always round the fondness to an integer rating). However, a user's fondness to any item can be modeled as a continuous variable in range(1, 5). Therefore, the "rating noise level" should be 0.5 - (e.g. fondness 3.5 and 4.5 can have the same rating = 4).

9. a human is not capable to accurately realize his fondness anyway (e.g. Alice: how much do you like this movie, 4.2 or 4.5? Bob: not sure.).

These 2 reasons seem to suggest high accuracy rating prediction is not theoretically possible at all. This is likely true for every single rating dataset, but unclear for joined rating datasets. I found a very good explanation on this issue from [Hinton's lecture](http://www.cs.toronto.edu/~bonner/courses/2014s/csc321/lectures/lec5.pdf). On slide 8, 3 highly correlated, organized, coarse-grain grids together provide fine-grain positioning for a point. Does this phenomenon exist in rating systems?

**Comment by lbholder** [comment]
If the current network can train to an arbitrarily low error, then it is probably too flexible and will overfit any dataset. And as you point out, there should always be some error on these datasets since there is inherent noise.

So, one approach would be to reduce the complexity of the network so that it no longer overfits the data, i.e., ideally converges to an error rate equal to the noise rate.

You can combine the above with a search over parameter space as well using a validation set (as you mentioned in our meeting, but just to be clear...):
1. Divide dataset into three sets: Train, Validate, Test.
2. Choose network parameters and train network using Train.
3. Test trained network on Validate set. Record error.
4. Repeat 2 and 3 until minimum error seems to have been achieved.
5. Evaluate this best network on the Test dataset, Record error.
6. Repeat 1-5 many times and average the errors from step 5 to get a final error for this approach.

**Comment by yuchenhou** [comment]
OK. I'll see how to prevent overfit and enhance generalization. Also I reviewed 2 papers on word embedding: [Distributed Representations of Words and Phrases and their Compositionality](http://papers.nips.cc/paper/5021-di) and [Efficient Estimation of Word Representations in Vector Space
](http://arxiv.org/abs/1301.3781) and payed attention to the configurations. Roughly, the configurations mentioned there have

### model configuration
```
{
  "embedding_size": 100 ~ 1000,
  "hidden_units_formation": 1 layer, 100~1000 units, or no hidden units at all
  "steps": billion ~ inf
}
```
### dataset specs
```
{
  "vocabulary": thousand ~ million
  "n_words": billion ~ inf
}
```
And they believe that shallow neural nets trade off small flexibility for big performance gain in embedding tasks. This should apply to our "node embedding" as well. Not sure if this applies to edge predictions.

**Comment by yuchenhou** [comment]
I've added early stopping and dropout to regularize model flexibility.
